src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/ContextAndRuleProviderState.cs

```diff
@@ -1,17 +1,18 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
 
 using Microsoft.Build.Framework.XamlTypes;
-using Microsoft.VisualStudio.ProjectSystem.Query.ProjectModel.Implementation;
 
 namespace Microsoft.VisualStudio.ProjectSystem.VS.Query
 {
     /// <summary>
-    /// Holds the state we need to pass from producers of <see cref="PropertyPageValue"/> instances
-    /// to other producers that will create the <see cref="PropertyPageValue"/>s' child entities.
+    /// Hold a <see cref="Rule"/> and the context in which to bind it. Used by 
+    /// <see cref="LaunchProfileDataProducer"/> and <see cref="PropertyPageDataProducer"/>
+    /// to pass the state their child producers will need, but allows the actual binding
+    /// of the <see cref="Rule"/> to be delayed until needed.
     /// </summary>
-    internal sealed class PropertyPageProviderState
+    internal sealed class ContextAndRuleProviderState
     {
-        public PropertyPageProviderState(IPropertyPageQueryCache cache, QueryProjectPropertiesContext context, Rule rule)
+        public ContextAndRuleProviderState(IPropertyPageQueryCache cache, QueryProjectPropertiesContext context, Rule rule)
```

Comment:
Similarly, wouldn't `RuleContextProviderState` work just as well?

Reply:
The `*ProviderState` types are really just property bags holding all the data we may need or want to pass to some dependent data producer. Naming these types is hard, because they don't really have a meaning or purpose beyond being a container.

For a while I didn't bother with types; I just passed tuples around. Then I started running into problems when I changed the shape of the tuple created by the producer, but not the shape expected by the consumer. Explicit types prevent that kind of problem. But naming them is hard. :-)